File: src/problems/015. 3Sum/Solution.py
# 15. 3Sum
# Accepted 2468ms

class Solution:
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ret = dict()
        
        nums.sort()
        for i, num_i in enumerate(nums):
            p, q = i + 1, len(nums) - 1
            while p < q:
                curVal = num_i + nums[p] + nums[q]
                if curVal < 0:
                    p += 1
                elif curVal > 0:
                    q -= 1
                else:
                    ret[str(num_i) + str(nums[p]) + str(nums[q])] = [num_i, nums[p], nums[q]]
                    p += 1; q -= 1
        
        return list(ret.values())

File: src/problems/015. 3Sum/Solution.js
// 15. 3Sum
// Accepted 1016ms

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    // -4, -1, -1, 0, 1, 2
    nums.sort((a, b) => a - b);
    
    let map = new Map();
    let ret = Array();
    
    for(let idx in nums) {
        let [p, q] = [parseInt(idx) + 1, nums.length - 1];
        
        while(p < q) {
            let sum = nums[idx] + nums[p] + nums[q];
            
            if(sum > 0)
                --q;
            else if(sum < 0)
                ++p;
            else {
                map.set(nums[idx].toString() + nums[p].toString() + nums[q].toString(), Array.of(nums[idx], nums[p], nums[q]));
                ++p;
                --q;
            }
        }
    }
    
    for(let [key, val] of map)
        ret.push(val);
    
    return ret;
};
